ord_schema/validations.py

```diff
@@ -0,0 +1,428 @@
+    if message.type == message.PressureControl.CUSTOM:
+        if not message.details:
+            raise ValueError('Pressure control custom, but no details provided')
+    if message.atmosphere == message.Atmosphere.CUSTOM:
+        if not message.atmosphere_details:
+            raise ValueError('Atmosphere custom, but no atmosphere_details provided')
+
+@return_message_if_valid
+def ValidatePressureMeasurement(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateStirringConditions(message):
+    ensure_float_nonnegative(message, 'rpm')
+    if message.type == message.StirringMethod.CUSTOM:
+        if not message.details:
+            raise ValueError('Stirring method custom, but no details provided')
+
+@return_message_if_valid
+def ValidateIlluminationConditions(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateElectrochemistryConditions(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateElectrochemistryMeasurement(message):
+    pass
+
+@return_message_if_valid
+def ValidateFlowConditions(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateTubing(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateReactionNotes(message):
+    pass
+
+@return_message_if_valid
+def ValidateReactionObservation(message):
+    pass
+
+@return_message_if_valid
+def ValidateReactionWorkup(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateReactionOutcome(message):
+    # Can only have one desired product
+    if sum(product.is_desired_product for product in message.products) > 1:
+        raise ValueError('Cannot have more than one desired product!')
+    # Check key values for product analyses
+    # NOTE(ccoley): Could use any(), but using expanded loops for clarity
+    analysis_keys = list(message.analyses.keys())
+    for product in message.products:
+        for field in ['analysis_identity', 'analysis_yield', 'analysis_purity',
+                      'analysis_selectivity']:
+            for key in getattr(product, field):
+                if key not in analysis_keys:
+                    raise ValueError(f'Undefined analysis key {key} '\
+                        'in ReactionProduct')
+
+@return_message_if_valid
+def ValidateReactionProduct(message):
+    if message.texture == message.Texture.CUSTOM:
+        if not message.texture_details:
+            raise ValueError(f'Custom texture defined for {type(message)}, ' \
+                'but texture_details field is empty')
+
+@return_message_if_valid
+def ValidateSelectivity(message):
+    ensure_float_nonnegative(message, 'precision')
+    if message.type == message.EE:
+        ensure_float_range(message, 'value', 0, 100)
+        if message.value > 0 and message.value < 1:
```

Comment:
Optional: you can do `if 0 < message.value < 1`

Reply:
For some reason I didn't think Python supported that, thanks